// @taquito/rpc/dist/taquito-rpc.es6.js
import { HttpBackend, HttpResponseError, STATUS_CODE } from '@taquito/http-utils';
import BigNumber from 'bignumber.js';
import { validateAddress, ValidationResult, InvalidAddressError, validateContractAddress, InvalidContractAddressError } from '@taquito/utils';

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */

function __rest(s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
}

function __awaiter(thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
}

const defaultChain = 'main';
const defaultRPCOptions = { block: 'head' };

/**
 * Casts object/array items to BigNumber
 * @param data input object or array
 * @param keys keys for processing or all items if not defined
 *
 */
function castToBigNumber(data, keys) {
    const returnArray = Array.isArray(data);
    if (typeof keys === 'undefined') {
        keys = Object.keys(data);
    }
    const response = returnArray ? [] : {};
    keys.forEach((key) => {
        const item = data[key];
        let res;
        if (typeof item === 'undefined') {
            return;
        }
        if (Array.isArray(item)) {
            res = castToBigNumber(item);
            response[key] = res;
            return;
        }
        res = new BigNumber(item);
        response[key] = res;
    });
    return response;
}

const defaultTtl = 1000;
/***
 * @description RpcClientCache acts as a decorator over the RpcClient instance by caching responses for the period defined by the ttl.
 */
class RpcClientCache {
    /**
     *
     * @param rpcClient rpcClient responsible of the interaction with Tezos network through an rpc node
     * @param ttl number representing the time to live (default 1000 milliseconds)
     *
     * @example new RpcClientCache(new RpcClient('https://mainnet.api.tez.ie/'))
     */
    constructor(rpcClient, ttl = defaultTtl) {
        this.rpcClient = rpcClient;
        this.ttl = ttl;
        this._cache = {};
    }
    getAllCachedData() {
        return this._cache;
    }
    /**
     * @description Remove all the data in the cache.
     *
     */
    deleteAllCachedData() {
        for (const key in this._cache) {
            delete this._cache[key];
        }
    }
    formatCacheKey(rpcUrl, rpcMethodName, rpcMethodParams, rpcMethodData) {
        let paramsToString = '';
        rpcMethodParams.forEach((param) => {
            paramsToString =
                typeof param === 'object'
                    ? paramsToString + JSON.stringify(param) + '/'
                    : paramsToString + param + '/';
        });
        return rpcMethodData
            ? `${rpcUrl}/${rpcMethodName}/${paramsToString}/${JSON.stringify(rpcMethodData)}`
            : `${rpcUrl}/${rpcMethodName}/${paramsToString}`;
    }
    has(key) {
        return key in this._cache;
    }
    get(key) {
        return this._cache[key].response;
    }
    put(key, response) {
        const handle = setTimeout(() => {
            return this.remove(key);
        }, this.ttl);
        Object.assign(this._cache, { [key]: { handle, response } });
    }
    remove(key) {
        if (key in this._cache) {
            delete this._cache[key];
        }
    }
    validateAddress(address) {
        if (validateAddress(address) !== ValidationResult.VALID) {
            throw new InvalidAddressError(`Invalid address: ${address}`);
        }
    }
    validateContract(address) {
        if (validateContractAddress(address) !== ValidationResult.VALID) {
            throw new InvalidContractAddressError(`Invalid contract address: ${address}`);
        }
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description Get the block's hash, its unique identifier.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-hash
     */
    getBlockHash({ block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getBlockHash', [block]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getBlockHash({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description List the ancestors of the given block which, if referred to as the branch in an operation header, are recent enough for that operation to be included in the current block.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-live-blocks
     */
    getLiveBlocks({ block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getLiveBlocks', [block]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getLiveBlocks({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param address address from which we want to retrieve the balance
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the balance of a contract.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-contracts-contract-id-balance
     */
    getBalance(address, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateAddress(address);
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getBalance', [block, address]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getBalance(address, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param address contract address from which we want to retrieve the storage
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the data of the contract.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-contracts-contract-id-storage
     */
    getStorage(address, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateContract(address);
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getStorage', [block, address]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getStorage(address, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param address contract address from which we want to retrieve the script
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the code and data of the contract.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-contracts-contract-id-script
     */
    getScript(address, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateContract(address);
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getScript', [block, address]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getScript(address, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param address contract address from which we want to retrieve the script
     * @param unparsingMode default is { unparsing_mode: "Readable" }
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the script of the contract and normalize it using the requested unparsing mode.
     *
     */
    getNormalizedScript(address, unparsingMode = { unparsing_mode: 'Readable' }, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateContract(address);
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getNormalizedScript', [
                block,
                address,
                unparsingMode,
            ]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getNormalizedScript(address, unparsingMode, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param address contract address from which we want to retrieve
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the complete status of a contract.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-contracts-contract-id
     */
    getContract(address, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateAddress(address);
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getContract', [block, address]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getContract(address, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param address contract address from which we want to retrieve the manager
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the manager key of a contract.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-contracts-contract-id-manager-key
     */
    getManagerKey(address, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateAddress(address);
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getManagerKey', [block, address]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getManagerKey(address, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param address contract address from which we want to retrieve the delegate (baker)
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the delegate of a contract, if any.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-contracts-contract-id-delegate
     */
    getDelegate(address, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateAddress(address);
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getDelegate', [block, address]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getDelegate(address, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param address contract address from which we want to retrieve the big map key
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the value associated with a key in the big map storage of the contract.
     *
     * @deprecated Deprecated in favor of getBigMapKeyByID
     *
     * @see https://tezos.gitlab.io/api/rpc.html#post-block-id-context-contracts-contract-id-big-map-get
     */
    getBigMapKey(address, key, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateAddress(address);
            const keyUrl = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getBigMapKey', [
                block,
                address,
                key,
            ]);
            if (this.has(keyUrl)) {
                return this.get(keyUrl);
            }
            else {
                const response = this.rpcClient.getBigMapKey(address, key, { block });
                this.put(keyUrl, response);
                return response;
            }
        });
    }
    /**
     *
     * @param id Big Map ID
     * @param expr Expression hash to query (A b58check encoded Blake2b hash of the expression (The expression can be packed using the pack_data method))
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the value associated with a key in a big map.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-big-maps-big-map-id-script-expr
     */
    getBigMapExpr(id, expr, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getBigMapExpr', [block, id, expr]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getBigMapExpr(id, expr, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param address delegate address which we want to retrieve
     * @param options contains generic configuration for rpc calls
     *
     * @description Fetches information about a delegate from RPC.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-delegates-pkh
     */
    getDelegates(address, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateAddress(address);
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getDelegates', [block, address]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getDelegates(address, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description All constants
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-constants
     */
    getConstants({ block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getConstants', [block]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getConstants({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls. See examples for various available sytaxes.
     *
     * @description All the information about a block
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id
     * @example getBlock() will default to /main/chains/block/head.
     * @example getBlock({ block: head~2 }) will return an offset of 2 blocks.
     * @example getBlock({ block: BL8fTiWcSxWCjiMVnDkbh6EuhqVPZzgWheJ2dqwrxYRm9AephXh~2 }) will return an offset of 2 blocks from given block hash..
     */
    getBlock({ block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getBlock', [block]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getBlock({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description The whole block header
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-header
     */
    getBlockHeader({ block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getBlockHeader', [block]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getBlockHeader({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description All the metadata associated to the block
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-metadata
     */
    getBlockMetadata({ block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getBlockMetadata', [block]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getBlockMetadata({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param args contains optional query arguments
     * @param options contains generic configuration for rpc calls
     *
     * @description Retrieves the list of delegates allowed to bake a block.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-helpers-baking-rights
     */
    getBakingRights(args = {}, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getBakingRights', [block, args]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getBakingRights(args, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param args contains optional query arguments
     * @param options contains generic configuration for rpc calls
     *
     * @description Retrieves the list of delegates allowed to bake a block.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-helpers-endorsing-rights
     */
    getEndorsingRights(args = {}, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getEndorsingRights', [
                block,
                args,
            ]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getEndorsingRights(args, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     * @param options contains generic configuration for rpc calls
     *
     * @description Ballots casted so far during a voting period
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-votes-ballot-list
     */
    getBallotList({ block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getBallotList', [block]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getBallotList({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description Sum of ballots casted so far during a voting period.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-votes-ballots
     */
    getBallots({ block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getBallots', [block]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getBallots({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description Current period kind.
     *
     * @deprecated Deprecated in favor of getCurrentPeriod
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-votes-current-period-kind
     */
    getCurrentPeriodKind({ block, } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getCurrentPeriodKind', [block]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getCurrentPeriodKind({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description Current proposal under evaluation.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-votes-current-proposal
     */
    getCurrentProposal({ block, } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getCurrentProposal', [block]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getCurrentProposal({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description Current expected quorum.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-votes-current-quorum
     */
    getCurrentQuorum({ block, } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getCurrentQuorum', [block]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getCurrentQuorum({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description List of delegates with their voting weight, in number of rolls.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-votes-listings
     */
    getVotesListings({ block, } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getVotesListings', [block]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getVotesListings({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description List of proposals with number of supporters.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-votes-proposals
     */
    getProposals({ block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getProposals', [block]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getProposals({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param data operation contents to forge
     * @param options contains generic configuration for rpc calls
     *
     * @description Forge an operation returning the unsigned bytes
     *
     * @see https://tezos.gitlab.io/api/rpc.html#post-block-id-helpers-forge-operations
     */
    forgeOperations(data, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.rpcClient.forgeOperations(data, { block });
        });
    }
    /**
     *
     * @param signedOpBytes signed bytes to inject
     *
     * @description Inject an operation in node and broadcast it. Returns the ID of the operation. The `signedOperationContents` should be constructed using a contextual RPCs from the latest block and signed by the client. By default, the RPC will wait for the operation to be (pre-)validated before answering. See RPCs under /blocks/prevalidation for more details on the prevalidation context.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#post-injection-operation
     */
    injectOperation(signedOpBytes) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.rpcClient.injectOperation(signedOpBytes);
        });
    }
    /**
     *
     * @param ops Operations to apply
     * @param options contains generic configuration for rpc calls
     *
     * @description Simulate the validation of an operation
     *
     * @see https://tezos.gitlab.io/api/rpc.html#post-block-id-helpers-preapply-operations
     */
    preapplyOperations(ops, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.rpcClient.preapplyOperations(ops, { block });
        });
    }
    /**
     *
     * @param contract address of the contract we want to get the entrypoints of
     *
     * @description Return the list of entrypoints of the contract
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-contracts-contract-id-entrypoints
     *
     * @version 005_PsBABY5H
     */
    getEntrypoints(contract, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateContract(contract);
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getEntrypoints', [
                block,
                contract,
            ]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getEntrypoints(contract, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     * @param op Operation to run
     * @param options contains generic configuration for rpc calls
     *
     * @description Run an operation without signature checks
     *
     * @see https://tezos.gitlab.io/api/rpc.html#post-block-id-helpers-scripts-run-operation
     */
    runOperation(op, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.rpcClient.runOperation(op, { block });
        });
    }
    /**
     * @param code Code to run
     * @param options contains generic configuration for rpc calls
     *
     * @description Run a piece of code in the current context
     *
     * @see https://tezos.gitlab.io/api/rpc.html#post-block-id-helpers-scripts-run-code
     */
    runCode(code, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.rpcClient.runCode(code, { block });
        });
    }
    getChainId() {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getChainId', []);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getChainId();
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param data Data to pack
     * @param options contains generic configuration for rpc calls
     *
     * @description Computes the serialized version of a data expression using the same algorithm as script instruction PACK
     *
     * @example packData({ data: { string: "test" }, type: { prim: "string" } })
     *
     * @see https://tezos.gitlab.io/api/rpc.html#post-block-id-helpers-scripts-pack-data
     */
    packData(data, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'packData', [block, data]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.packData(data, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @description Return rpc root url
     */
    getRpcUrl() {
        return this.rpcClient.getRpcUrl();
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description Voting period of current block.
     *
     * @example getCurrentPeriod() will default to current voting period for /main/chains/block/head.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-votes-current-period
     */
    getCurrentPeriod({ block, } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getCurrentPeriod', [block]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getCurrentPeriod({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description Voting period of next block.
     *
     * @example getSuccessorPeriod() will default to successor voting period for /main/chains/block/head.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-votes-successor-period
     */
    getSuccessorPeriod({ block, } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getSuccessorPeriod', [block]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getSuccessorPeriod({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param id Sapling state ID
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the value associated with a sapling state ID.
     *
     * @see https://tezos.gitlab.io/active/rpc.html#get-block-id-context-sapling-sapling-state-id-get-diff
     */
    getSaplingDiffById(id, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getSaplingDiffById', [block, id]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getSaplingDiffById(id, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param contract address of the contract we want to get the sapling diff
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the value associated with a sapling state.
     *
     * @see https://tezos.gitlab.io/active/rpc.html#get-block-id-context-contracts-contract-id-single-sapling-get-diff
     */
    getSaplingDiffByContract(contract, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), 'getSaplingDiffByContract', [
                block,
                contract,
            ]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getSaplingDiffByContract(contract, { block });
                this.put(key, response);
                return response;
            }
        });
    }
}

var OpKind;
(function (OpKind) {
    OpKind["ORIGINATION"] = "origination";
    OpKind["DELEGATION"] = "delegation";
    OpKind["REVEAL"] = "reveal";
    OpKind["TRANSACTION"] = "transaction";
    OpKind["ACTIVATION"] = "activate_account";
    OpKind["ENDORSEMENT"] = "endorsement";
    OpKind["ENDORSEMENT_WITH_SLOT"] = "endorsement_with_slot";
    OpKind["SEED_NONCE_REVELATION"] = "seed_nonce_revelation";
    OpKind["DOUBLE_ENDORSEMENT_EVIDENCE"] = "double_endorsement_evidence";
    OpKind["DOUBLE_BAKING_EVIDENCE"] = "double_baking_evidence";
    OpKind["PROPOSALS"] = "proposals";
    OpKind["BALLOT"] = "ballot";
    OpKind["FAILING_NOOP"] = "failing_noop";
    OpKind["REGISTER_GLOBAL_CONSTANT"] = "register_global_constant";
})(OpKind || (OpKind = {}));

// IMPORTANT: THIS FILE IS AUTO GENERATED! DO NOT MANUALLY EDIT OR CHECKIN!
const VERSION = {
    "commitHash": "81f0a5b103f867f57fbe5d526315c375a3788346",
    "version": "11.2.0"
};

/***
 * @description RpcClient allows interaction with Tezos network through an rpc node
 */
class RpcClient {
    /**
     *
     * @param url rpc root url
     * @param chain chain (default main)
     * @param httpBackend Http backend that issue http request.
     * You can override it by providing your own if you which to hook in the request/response
     *
     * @example new RpcClient('https://mainnet.api.tez.ie/', 'main') this will use https://mainnet.api.tez.ie//chains/main
     */
    constructor(url, chain = defaultChain, httpBackend = new HttpBackend()) {
        this.url = url;
        this.chain = chain;
        this.httpBackend = httpBackend;
    }
    createURL(path) {
        // Trim trailing slashes because it is assumed to be included in path
        return `${this.url.replace(/\/+$/g, '')}${path}`;
    }
    validateAddress(address) {
        if (validateAddress(address) !== ValidationResult.VALID) {
            throw new InvalidAddressError(`Invalid address: ${address}`);
        }
    }
    validateContract(address) {
        if (validateContractAddress(address) !== ValidationResult.VALID) {
            throw new InvalidAddressError(`Invalid address: ${address}`);
        }
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description Get the block's hash, its unique identifier.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-hash
     */
    getBlockHash({ block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const hash = yield this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/hash`),
                method: 'GET',
            });
            return hash;
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description List the ancestors of the given block which, if referred to as the branch in an operation header, are recent enough for that operation to be included in the current block.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-live-blocks
     */
    getLiveBlocks({ block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const blocks = yield this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/live_blocks`),
                method: 'GET',
            });
            return blocks;
        });
    }
    /**
     *
     * @param address address from which we want to retrieve the balance
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the balance of a contract.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-contracts-contract-id-balance
     */
    getBalance(address, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateAddress(address);
            const balance = yield this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/context/contracts/${address}/balance`),
                method: 'GET',
            });
            return new BigNumber(balance);
        });
    }
    /**
     *
     * @param address contract address from which we want to retrieve the storage
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the data of the contract.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-contracts-contract-id-storage
     */
    getStorage(address, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateContract(address);
            return this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/context/contracts/${address}/storage`),
                method: 'GET',
            });
        });
    }
    /**
     *
     * @param address contract address from which we want to retrieve the script
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the code and data of the contract.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-contracts-contract-id-script
     */
    getScript(address, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateContract(address);
            return this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/context/contracts/${address}/script`),
                method: 'GET',
            });
        });
    }
    /**
     *
     * @param address contract address from which we want to retrieve the script
     * @param unparsingMode default is { unparsing_mode: "Readable" }
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the script of the contract and normalize it using the requested unparsing mode.
     *
     */
    getNormalizedScript(address, unparsingMode = { unparsing_mode: 'Readable' }, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateContract(address);
            return this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/context/contracts/${address}/script/normalized`),
                method: 'POST',
            }, unparsingMode);
        });
    }
    /**
     *
     * @param address contract address from which we want to retrieve
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the complete status of a contract.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-contracts-contract-id
     */
    getContract(address, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateAddress(address);
            const contractResponse = yield this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/context/contracts/${address}`),
                method: 'GET',
            });
            return Object.assign(Object.assign({}, contractResponse), { balance: new BigNumber(contractResponse.balance) });
        });
    }
    /**
     *
     * @param address contract address from which we want to retrieve the manager
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the manager key of a contract.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-contracts-contract-id-manager-key
     */
    getManagerKey(address, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateAddress(address);
            return this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/context/contracts/${address}/manager_key`),
                method: 'GET',
            });
        });
    }
    /**
     *
     * @param address contract address from which we want to retrieve the delegate (baker)
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the delegate of a contract, if any.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-contracts-contract-id-delegate
     */
    getDelegate(address, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateAddress(address);
            let delegate;
            try {
                delegate = yield this.httpBackend.createRequest({
                    url: this.createURL(`/chains/${this.chain}/blocks/${block}/context/contracts/${address}/delegate`),
                    method: 'GET',
                });
            }
            catch (ex) {
                if (ex instanceof HttpResponseError && ex.status === STATUS_CODE.NOT_FOUND) {
                    delegate = null;
                }
                else {
                    throw ex;
                }
            }
            return delegate;
        });
    }
    /**
     *
     * @param address contract address from which we want to retrieve the big map key
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the value associated with a key in the big map storage of the contract.
     *
     * @deprecated Deprecated in favor of getBigMapKeyByID
     *
     * @see https://tezos.gitlab.io/api/rpc.html#post-block-id-context-contracts-contract-id-big-map-get
     */
    getBigMapKey(address, key, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateAddress(address);
            return this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/context/contracts/${address}/big_map_get`),
                method: 'POST',
            }, key);
        });
    }
    /**
     *
     * @param id Big Map ID
     * @param expr Expression hash to query (A b58check encoded Blake2b hash of the expression (The expression can be packed using the pack_data method))
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the value associated with a key in a big map.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-big-maps-big-map-id-script-expr
     */
    getBigMapExpr(id, expr, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/context/big_maps/${id}/${expr}`),
                method: 'GET',
            });
        });
    }
    /**
     *
     * @param address delegate address which we want to retrieve
     * @param options contains generic configuration for rpc calls
     *
     * @description Fetches information about a delegate from RPC.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-delegates-pkh
     */
    getDelegates(address, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateAddress(address);
            const response = yield this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/context/delegates/${address}`),
                method: 'GET',
            });
            return {
                deactivated: response.deactivated,
                balance: new BigNumber(response.balance),
                frozen_balance: new BigNumber(response.frozen_balance),
                frozen_balance_by_cycle: response.frozen_balance_by_cycle.map((_a) => {
                    var { deposit, deposits, fees, rewards } = _a, rest = __rest(_a, ["deposit", "deposits", "fees", "rewards"]);
                    const castedToBigNumber = castToBigNumber({ deposit, deposits, fees, rewards }, [
                        'deposit',
                        'deposits',
                        'fees',
                        'rewards',
                    ]);
                    return Object.assign(Object.assign({}, rest), { deposit: castedToBigNumber.deposit, deposits: castedToBigNumber.deposits, fees: castedToBigNumber.fees, rewards: castedToBigNumber.rewards });
                }),
                staking_balance: new BigNumber(response.staking_balance),
                delegated_contracts: response.delegated_contracts,
                delegated_balance: new BigNumber(response.delegated_balance),
                grace_period: response.grace_period,
                voting_power: response.voting_power,
            };
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description All constants
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-constants
     */
    getConstants({ block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/context/constants`),
                method: 'GET',
            });
            const castedResponse = castToBigNumber(response, [
                'time_between_blocks',
                'hard_gas_limit_per_operation',
                'hard_gas_limit_per_block',
                'proof_of_work_threshold',
                'tokens_per_roll',
                'seed_nonce_revelation_tip',
                'block_security_deposit',
                'endorsement_security_deposit',
                'block_reward',
                'endorsement_reward',
                'cost_per_byte',
                'hard_storage_limit_per_operation',
                'test_chain_duration',
                'baking_reward_per_endorsement',
                'delay_per_missing_endorsement',
                'minimal_block_delay',
                'liquidity_baking_subsidy',
                'cache_layout',
                'baking_reward_fixed_portion',
                'baking_reward_bonus_per_slot',
                'endorsing_reward_per_slot',
                'double_baking_punishment',
                'delay_increment_per_round',
            ]);
            return Object.assign(Object.assign({}, response), castedResponse);
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls. See examples for various available sytaxes.
     *
     * @description All the information about a block
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id
     * @example getBlock() will default to /main/chains/block/head.
     * @example getBlock({ block: head~2 }) will return an offset of 2 blocks.
     * @example getBlock({ block: BL8fTiWcSxWCjiMVnDkbh6EuhqVPZzgWheJ2dqwrxYRm9AephXh~2 }) will return an offset of 2 blocks from given block hash..
     */
    getBlock({ block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}`),
                method: 'GET',
            });
            return response;
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description The whole block header
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-header
     */
    getBlockHeader({ block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/header`),
                method: 'GET',
            });
            return response;
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description All the metadata associated to the block
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-metadata
     */
    getBlockMetadata({ block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/metadata`),
                method: 'GET',
            });
            return response;
        });
    }
    /**
     *
     * @param args contains optional query arguments
     * @param options contains generic configuration for rpc calls
     *
     * @description Retrieves the list of delegates allowed to bake a block.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-helpers-baking-rights
     */
    getBakingRights(args = {}, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/helpers/baking_rights`),
                method: 'GET',
                query: args,
            });
            return response;
        });
    }
    /**
     *
     * @param args contains optional query arguments
     * @param options contains generic configuration for rpc calls
     *
     * @description Retrieves the list of delegates allowed to bake a block.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-helpers-endorsing-rights
     */
    getEndorsingRights(args = {}, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/helpers/endorsing_rights`),
                method: 'GET',
                query: args,
            });
            return response;
        });
    }
    /**
     * @param options contains generic configuration for rpc calls
     *
     * @description Ballots casted so far during a voting period
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-votes-ballot-list
     */
    getBallotList({ block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/votes/ballot_list`),
                method: 'GET',
            });
            return response;
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description Sum of ballots casted so far during a voting period.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-votes-ballots
     */
    getBallots({ block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/votes/ballots`),
                method: 'GET',
            });
            return response;
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description Current period kind.
     *
     * @deprecated Deprecated in favor of getCurrentPeriod
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-votes-current-period-kind
     */
    getCurrentPeriodKind({ block, } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/votes/current_period_kind`),
                method: 'GET',
            });
            return response;
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description Current proposal under evaluation.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-votes-current-proposal
     */
    getCurrentProposal({ block, } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/votes/current_proposal`),
                method: 'GET',
            });
            return response;
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description Current expected quorum.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-votes-current-quorum
     */
    getCurrentQuorum({ block, } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/votes/current_quorum`),
                method: 'GET',
            });
            return response;
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description List of delegates with their voting weight, in number of rolls.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-votes-listings
     */
    getVotesListings({ block, } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/votes/listings`),
                method: 'GET',
            });
            return response;
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description List of proposals with number of supporters.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-votes-proposals
     */
    getProposals({ block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/votes/proposals`),
                method: 'GET',
            });
            return response;
        });
    }
    /**
     *
     * @param data operation contents to forge
     * @param options contains generic configuration for rpc calls
     *
     * @description Forge an operation returning the unsigned bytes
     *
     * @see https://tezos.gitlab.io/api/rpc.html#post-block-id-helpers-forge-operations
     */
    forgeOperations(data, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/helpers/forge/operations`),
                method: 'POST',
            }, data);
        });
    }
    /**
     *
     * @param signedOpBytes signed bytes to inject
     *
     * @description Inject an operation in node and broadcast it. Returns the ID of the operation. The `signedOperationContents` should be constructed using a contextual RPCs from the latest block and signed by the client. By default, the RPC will wait for the operation to be (pre-)validated before answering. See RPCs under /blocks/prevalidation for more details on the prevalidation context.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#post-injection-operation
     */
    injectOperation(signedOpBytes) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.httpBackend.createRequest({
                url: this.createURL(`/injection/operation`),
                method: 'POST',
            }, signedOpBytes);
        });
    }
    /**
     *
     * @param ops Operations to apply
     * @param options contains generic configuration for rpc calls
     *
     * @description Simulate the validation of an operation
     *
     * @see https://tezos.gitlab.io/api/rpc.html#post-block-id-helpers-preapply-operations
     */
    preapplyOperations(ops, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/helpers/preapply/operations`),
                method: 'POST',
            }, ops);
            return response;
        });
    }
    /**
     *
     * @param contract address of the contract we want to get the entrypoints of
     *
     * @description Return the list of entrypoints of the contract
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-contracts-contract-id-entrypoints
     *
     * @version 005_PsBABY5H
     */
    getEntrypoints(contract, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateContract(contract);
            const contractResponse = yield this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/context/contracts/${contract}/entrypoints`),
                method: 'GET',
            });
            return contractResponse;
        });
    }
    /**
     * @param op Operation to run
     * @param options contains generic configuration for rpc calls
     *
     * @description Run an operation without signature checks
     *
     * @see https://tezos.gitlab.io/api/rpc.html#post-block-id-helpers-scripts-run-operation
     */
    runOperation(op, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/helpers/scripts/run_operation`),
                method: 'POST',
            }, op);
            return response;
        });
    }
    /**
     * @param code Code to run
     * @param options contains generic configuration for rpc calls
     *
     * @description Run a piece of code in the current context
     *
     * @see https://tezos.gitlab.io/api/rpc.html#post-block-id-helpers-scripts-run-code
     */
    runCode(code, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/helpers/scripts/run_code`),
                method: 'POST',
            }, code);
            return response;
        });
    }
    getChainId() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/chain_id`),
                method: 'GET',
            });
        });
    }
    /**
     *
     * @param data Data to pack
     * @param options contains generic configuration for rpc calls
     *
     * @description Computes the serialized version of a data expression using the same algorithm as script instruction PACK
     * Note: You should always verify the packed bytes before signing or requesting that they be signed when using the the RPC to pack.
     * This precaution helps protect you and your applications users from RPC nodes that have been compromised.
     * A node that is operated by a bad actor, or compromised by a bad actor could return a fully formed operation that does not correspond to the input provided to the RPC endpoint.
     * A safer solution to pack and sign data would be to use the `packDataBytes` function available in the `@taquito/michel-codec` package.
     *
     * @example packData({ data: { string: "test" }, type: { prim: "string" } })
     *
     *
     * @see https://tezos.gitlab.io/api/rpc.html#post-block-id-helpers-scripts-pack-data
     */
    packData(data, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const _a = yield this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/helpers/scripts/pack_data`),
                method: 'POST',
            }, data), { gas } = _a, rest = __rest(_a, ["gas"]);
            let formattedGas = gas;
            const tryBigNumber = new BigNumber(gas || '');
            if (!tryBigNumber.isNaN()) {
                formattedGas = tryBigNumber;
            }
            return Object.assign({ gas: formattedGas }, rest);
        });
    }
    /**
     *
     * @description Return rpc root url
     */
    getRpcUrl() {
        return this.url;
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description Voting period of current block.
     *
     * @example getCurrentPeriod() will default to current voting period for /main/chains/block/head.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-votes-current-period
     */
    getCurrentPeriod({ block, } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/votes/current_period`),
                method: 'GET',
            });
            return response;
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description Voting period of next block.
     *
     * @example getSuccessorPeriod() will default to successor voting period for /main/chains/block/head.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-votes-successor-period
     */
    getSuccessorPeriod({ block, } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/votes/successor_period`),
                method: 'GET',
            });
            return response;
        });
    }
    /**
     *
     * @param id Sapling state ID
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the value associated with a sapling state ID.
     *
     * @see https://tezos.gitlab.io/active/rpc.html#get-block-id-context-sapling-sapling-state-id-get-diff
     */
    getSaplingDiffById(id, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/context/sapling/${id}/get_diff`),
                method: 'GET',
            });
        });
    }
    /**
     *
     * @param contract address of the contract we want to get the sapling diff
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the value associated with a sapling state.
     *
     * @see https://tezos.gitlab.io/active/rpc.html#get-block-id-context-contracts-contract-id-single-sapling-get-diff
     */
    getSaplingDiffByContract(contract, { block } = defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.httpBackend.createRequest({
                url: this.createURL(`/chains/${this.chain}/blocks/${block}/context/contracts/${contract}/single_sapling_get_diff`),
                method: 'GET',
            });
        });
    }
}

export { OpKind, RpcClient, RpcClientCache, VERSION, castToBigNumber, defaultChain, defaultRPCOptions };
