// @taquito/michelson-encoder/dist/lib/schema/parameter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ParameterSchema = void 0;
const createToken_1 = require("../tokens/createToken");
const token_1 = require("../tokens/token");
const or_1 = require("../tokens/or");
const option_1 = require("../tokens/option");
/**
 * @warn Our current smart contract abstraction feature is currently in preview. It's API is not final, and it may not cover every use case (yet). We will greatly appreciate any feedback on this feature.
 */
class ParameterSchema {
    constructor(val) {
        this.root = createToken_1.createToken(val, 0);
    }
    static fromRPCResponse(val) {
        const parameter = val &&
            val.script &&
            Array.isArray(val.script.code) &&
            val.script.code.find((x) => x.prim === 'parameter');
        if (!parameter || !Array.isArray(parameter.args)) {
            throw new Error('Invalid rpc response passed as arguments');
        }
        return new ParameterSchema(parameter.args[0]);
    }
    get isMultipleEntryPoint() {
        return (this.root instanceof or_1.OrToken ||
            (this.root instanceof option_1.OptionToken && this.root.subToken() instanceof or_1.OrToken));
    }
    get hasAnnotation() {
        if (this.isMultipleEntryPoint) {
            return Object.keys(this.ExtractSchema())[0] !== '0';
        }
        else {
            return true;
        }
    }
    Execute(val, semantics) {
        return this.root.Execute(val, semantics);
    }
    Encode(...args) {
        try {
            return this.root.Encode(args.reverse());
        }
        catch (ex) {
            if (ex instanceof token_1.TokenValidationError) {
                throw ex;
            }
            throw new Error(`Unable to encode parameter. ${ex}`);
        }
    }
    EncodeObject(_value) {
        try {
            return this.root.EncodeObject(_value);
        }
        catch (ex) {
            if (ex instanceof token_1.TokenValidationError) {
                throw ex;
            }
            throw new Error(`Unable to encode parameter object. ${ex}`);
        }
    }
    /**
     * @deprecated ExtractSchema has been deprecated in favor of generateSchema
     *
     */
    ExtractSchema() {
        return this.root.ExtractSchema();
    }
    /**
     * @description Produce a schema grouping together all the entry points of a contract.
     */
    generateSchema() {
        return this.root.generateSchema();
    }
    ExtractSignatures() {
        return this.root.ExtractSignature();
    }
}
exports.ParameterSchema = ParameterSchema;
